modules/component/src/listComponentState.ts

```diff
@@ -0,0 +1,44 @@
+import {Either, List} from 'standard-data-structures'
+
+type ListKey = string | number
+
+export class ListComponentState<S> {
+  private constructor(
+    private baseInit: () => S,
+    private readonly items: List<ListKey>,
+    private readonly lookUp: {[k in ListKey]?: S}
+  ) {}
+  static of<T>(baseInit: () => T) {
+    return new ListComponentState(baseInit, List.empty(), {})
+  }
+  set(k: ListKey, value: S = this.baseInit()) {
+    const itemState = this.lookUp[k]
+    if (itemState !== undefined) {
```

Comment:
What happens if we are creating a list from an empty component? I think we should be using `hasOwnProperty` or use a `Map`